### DISCUSSION TITLE
Migrating jwk_convert.py to Python 3

### DISCUSSION THREAD

**Discussion by szepeviktor** [candango/automatoes#114] [General]
@piraz Could you help @jsabater to convert his key?

https://github.com/szepeviktor/debian-server-tools/blob/master/security/jwk_convert.py

> Hi, Viktor!
> 
> Just tried to use your tool to convert a private_key.json from the latest version of Certbot to PEM format using Python 3.9 under Debian 11 Bullseye and found this error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/ansible/bin/jwk_convert.py", line 31, in <module>
>     PKEY[k] = enc(v.encode())
>   File "/home/ansible/bin/jwk_convert.py", line 26, in enc
>     return '0x'+binascii.hexlify(base64.b64decode(data, b'-_')).upper()
> TypeError: can only concatenate str (not "bytes") to str
> ```
> 
> I had converted the script to Python 3 using `2to3` but handling strings of bytes seems to need specific knowledge, which I don't have. I think the problem is here:
> 
> ```
> data += b'=' * missing_padding
> ```
> 
> I've tried to research this topic without success. Would you feel like upgrading the script to Python 3? I think that the need to convert JWK to PEM is going to be there for a while for all us Ansible users.
> 
> Incidentally, I'd love to create an Ansible Playbook to do the whole process, maybe through a module.
> 
> Thanks in advance and happy new year.

**Comment by piraz** [comment]
I think that was fixed here #109. Need to release it.

I'll do it today.

**Comment by piraz** [comment]
@szepeviktor now I understand what is going on. Let me see if I can help.

By the way, 0.9.9 is working right now. Automatoes will work fine with python 3.11.

**Reply by piraz** [reply]
The error is on the data's format and at the concatenation from the return. If data is string, than you will have an error:

```python
def enc(data):
    missing_padding = 4 - len(data) % 4
    if missing_padding:
        data += b'=' * missing_padding #  <==== right here, because you are concatenating strings with bytes

    return '0x'+binascii.hexlify(base64.b64decode(data, b'-_')).upper()  # <=== and here, becuase '0x' is string and binascii.hexlify will return bytes
```

You can check if data is string and convert it to bytes and convert the '0x' to b'0x', which is the byte form.

Here is your function working:

```python
def enc(data):
    if isinstance(data, str):
        data = data.encode()
    missing_padding = 4 - len(data) % 4
    if missing_padding:
        data += b"=" * missing_padding
    return b"0x" + binascii.hexlify(base64.b64decode(data, b"-_")).upper()
```

I like double quotes for strings, you can change it if you want. ;)
